### PR TITLE
[#1045] fix(server): shuffle server may hang when restart worker due to multi require-momery and no require-momery release

### PR DESCRIPTION
**Pull request by lifeSo**
### What changes were proposed in this pull request?

shuffle server may hang when restart worker due to multi require-momery and no require-momery release.
So, fix the bug

### Why are the changes needed?

shuffle server may hang when restart worker due to multi require-momery and no require-momery release.
So, fix the bug

Fix:  https://github.com/apache/incubator-uniffle/issues/1045

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?
unit test

### CODE REVIEW

**Comment by jerqi** [comment]
@zuston Could you help me review this pr?

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 1]
```diff
@@ -408,6 +408,10 @@ public long getPartitionDataSize(String appId, int shuffleId, int partitionId) {
 
   public long requireBuffer(
       String appId, int shuffleId, List<Integer> partitionIds, int requireSize) {
+    boolean isRegistered = shuffleBufferManager.checkIfRegistered(appId, shuffleId, partitionIds);
+    if (!isRegistered) {
+      return -4; // Keep the same with StatusCode
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:413]: Could give `-1` and `-4` a error code? like
```
enum ERRORCODE {
NO_BUFFER(-1),
NO_REGISTER(-4)
}
```
Or we can reuse the error code which existed.

File: client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/impl/RssTezFetcherTask.java [hunk 2]
```diff
@@ -186,6 +187,7 @@ protected FetchResult callInternal() throws Exception {
               taskIdBitmap,
               new ArrayList<>(serverInfoSet),
               readerJobConf,
+              new TezIdHelper(),
```
Comment (jerqi) [inline client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/impl/RssTezFetcherTask.java:190]: Is this another bug? Will it cause correctness issues?

File: server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java [hunk 3]
```diff
@@ -189,6 +189,20 @@ private void updateShuffleSize(String appId, int shuffleId, long size) {
     shuffleIdToSize.get(shuffleId).addAndGet(size);
   }
 
+  public boolean checkIfRegistered(String appId, int shuffleId, List<Integer> partitionIds) {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java:192]: Could we use ` ShuffleTaskManager#shuffleTaskInfos`  to judge whether is registered?

Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/buffer/ShuffleBufferManager.java:192]: +1 I have similar thought. The appId + shuffleId check existence are enough. Right?

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java (hunk 1 above)
Comment (lifeSo) [reply server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java]: I find there is no error code existed I can use, and I craete RequireBufferStatusCode

**Comment by lifeSo** [comment]
How about the spark test code, I did not change the spark test.

**Comment by jerqi** [comment]
> How about the spark test code, I did not change the spark test.

But you modify the code shuffle server, Spark integration tests will use shuffle server.

**Comment by lifeSo** [comment]
> 

Ok, may be the spark test not register, and I only changed register logical.

**Comment by jerqi** [comment]
> > 
> 
> Ok, may be the spark test not register, and I only changed register logical.

Spark register the application in the RssShuffleManager.

**Comment by lifeSo** [comment]
> RssShuffleManager

In fact, likes not registered from my test and log

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 4]
```diff
@@ -409,14 +410,15 @@ public long getPartitionDataSize(String appId, int shuffleId, int partitionId) {
...
     ShuffleTaskInfo shuffleTaskInfo = shuffleTaskInfos.get(appId);
-    if (shuffleTaskInfo != null) {
-      for (int partitionId : partitionIds) {
-        long partitionUsedDataSize = getPartitionDataSize(appId, shuffleId, partitionId);
-        if (shuffleBufferManager.limitHugePartition(
-            appId, shuffleId, partitionId, partitionUsedDataSize)) {
-          ShuffleServerMetrics.counterTotalRequireBufferFailedForHugePartition.inc();
-          return -1;
-        }
+    if (null == shuffleTaskInfo && !"EMPTY".equals(appId)) {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:413]: Why do we add `EMPTY` here? It's weird.

**Review by jerqi — APPROVED** [review]
LGTM, thanks @lifeSo

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java [hunk 5]
```diff
@@ -409,14 +410,15 @@ public long getPartitionDataSize(String appId, int shuffleId, int partitionId) {
...
-        long partitionUsedDataSize = getPartitionDataSize(appId, shuffleId, partitionId);
-        if (shuffleBufferManager.limitHugePartition(
-            appId, shuffleId, partitionId, partitionUsedDataSize)) {
-          ShuffleServerMetrics.counterTotalRequireBufferFailedForHugePartition.inc();
-          return -1;
-        }
+    if (null == shuffleTaskInfo) {
+      return RequireBufferStatusCode.NO_REGISTER.statusCode();
+    }
+    for (int partitionId : partitionIds) {
```
Comment (jerqi) [inline server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:416]: @zuston Could you take another look at this place?

**Comment by jerqi** [comment]
@zuston Could you help me this pr?

File: client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/impl/RssTezFetcher.java [hunk 6]
```diff
@@ -133,7 +133,7 @@ public void copyFromRssServer() throws IOException {
         hasPendingData = false;
         uncompressedData = null;
       } else {
-        LOG.info("uncompressedData is null");
+        LOG.info("UncompressedData is null");
```
Comment (zuston) [inline client-tez/src/main/java/org/apache/tez/runtime/library/common/shuffle/impl/RssTezFetcher.java:136]: I think this is not related with current issue. 

Anyway, let's reserve these changes.

File: server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java (hunk 5 above)
Comment (zuston) [reply server/src/main/java/org/apache/uniffle/server/ShuffleTaskManager.java:416]: I think the change is right. But I will introduce new grpc status code like `BUFFER_LIMIT_FOR_HUGE_PARTITION` instead of `NO_BUFFER`

**Review by zuston — APPROVED** [review]
lgtm.
